File: compiler/src/main/java/edu/cornell/cs/apl/viaduct/algebra/FreeDistributiveLattice.java
package edu.cornell.cs.apl.viaduct.algebra;

import com.google.auto.value.AutoValue;
import io.vavr.collection.Array;
import io.vavr.collection.HashSet;
import io.vavr.collection.Set;
import java.util.function.Function;

/**
 * The free distributive lattice over an arbitrary set {@code A} of elements. In addition to lattice
 * identities, the following hold:
 *
 * <p>{@code a /\ (b \/ c) == (a /\ b) \/ (a /\ c)}
 *
 * <p>{@code a \/ (b /\ c) == (a \/ b) /\ (a \/ c)}
 */
@AutoValue
public abstract class FreeDistributiveLattice<A>
    implements HeytingAlgebra<FreeDistributiveLattice<A>> {
  private static final FreeDistributiveLattice<?> BOTTOM = create(HashSet.of());

  private static final FreeDistributiveLattice<?> TOP = create(HashSet.of(HashSet.of()));

  public static <A> FreeDistributiveLattice<A> create(A element) {
    return new AutoValue_FreeDistributiveLattice<>(HashSet.of(HashSet.of(element)));
  }

  private static <A> FreeDistributiveLattice<A> create(Set<Set<A>> joinOfMeets) {
    return new AutoValue_FreeDistributiveLattice<>(removeRedundant(joinOfMeets));
  }

  public static <A> FreeDistributiveLattice<A> top() {
    return (FreeDistributiveLattice<A>) TOP;
  }

  public static <A> FreeDistributiveLattice<A> bottom() {
    return (FreeDistributiveLattice<A>) BOTTOM;
  }

  /** Remove redundant meets according to {@link static isRedundant}. */
  private static <A> Set<Set<A>> removeRedundant(Set<Set<A>> joinOfMeets) {
    return joinOfMeets.filter((meet) -> !isRedundant(joinOfMeets, meet));
  }

  /**
   * Given {@code m_1 \/ m_2 \/ ... \/ m_n}, if any {@code m_i} is a strict subset of {@code m_j},
   * then {@code m_j} is redundant.
   */
  private static <A> boolean isRedundant(Set<Set<A>> joinOfMeets, Set<A> j) {
    for (Set<A> i : joinOfMeets) {
      if (!i.equals(j) && j.containsAll(i)) {
        return true;
      }
    }
    return false;
  }

  abstract Set<Set<A>> getJoinOfMeets();

  @Override
  public final boolean lessThanOrEqualTo(FreeDistributiveLattice<A> that) {
    return this.join(that).equals(that);
  }

  @Override
  public final FreeDistributiveLattice<A> join(FreeDistributiveLattice<A> with) {
    return create(this.getJoinOfMeets().union(with.getJoinOfMeets()));
  }

  @Override
  public final FreeDistributiveLattice<A> meet(FreeDistributiveLattice<A> with) {
    Set<Set<A>> candidates = HashSet.empty();

    for (Set<A> meet1 : this.getJoinOfMeets()) {
      for (Set<A> meet2 : with.getJoinOfMeets()) {
        candidates = candidates.add(meet1.union(meet2));
      }
    }

    return create(candidates);
  }

  /**
   * Returns the relative pseudocomplement of {@code this} relative to {@code that}. the relative
   * pseudocomplement is greatest x s.t. {@code this & x <= that}.
   *
   * <p>How does this work? we are dealing with constraints of the form {@code (A1 | ... | Am) & x
   * <= B1 | ... | Bn}
   *
   * <p>which can be rewritten as {@code (A1&x) | ... | (Am&x) <= B1 | ... | Bn}
   *
   * <p>This inequality only holds true if every meet on the left can be "covered" on the right s.t.
   * a meet on the right side is a subset of the meet in the left side. For every meet on the left
   * Ai, we complement it with every meet on the right Bj. because we want the greatest solution, we
   * join these complements together, arriving at an upper bound for x: {@code x <= Ci1 | ... | Cin}
   *
   * <p>where {@code Cij = Bj \ Ai}.
   *
   * <p>But we have to do the same process for all meets on the left, so we get m upper bounds.
   * these have to be all simultaneously satisfied, so we take the meet of the upper bounds: {@code
   * x = (C11 | ... | C1n) & ... & (Cm1 | ... | Cmn)}
   *
   * <p>The algorithm below computes exactly this solution.
   */
  @Override
  public final FreeDistributiveLattice<A> imply(FreeDistributiveLattice<A> that) {
    // TODO: make sure this is correct
    return this.getJoinOfMeets()
        .foldRight(
            top(),
            (thisMeet, acc) -> {
              Set<Set<A>> newJoinOfMeets =
                  that.getJoinOfMeets().map((thatMeet) -> thatMeet.removeAll(thisMeet));
              return acc.meet(create(newJoinOfMeets));
            });
  }

  @Override
  public String toString() {
    final Function<Set<A>, String> meetToString =
        (meet) -> {
          final Array<String> elements = meet.toArray().sorted().map(Object::toString);
          final String body = String.join(" \u2227 ", elements);
          return meet.length() > 1 ? "(" + body + ")" : body;
        };

    if (this.equals(top())) {
      return "\u22A4";
    } else if (this.equals(bottom())) {
      return "\u22A5";
    } else {
      final Array<String> meets = getJoinOfMeets().toArray().map(meetToString).sorted();
      final String body = String.join(" \u2228 ", meets);
      return getJoinOfMeets().length() > 1 ? "(" + body + ")" : body;
    }
  }
}

File: compiler/src/main/kotlin/io/github/apl_cornell/viaduct/algebra/FreeDistributiveLattice.kt
package io.github.apl_cornell.viaduct.algebra

import kotlinx.collections.immutable.PersistentSet
import kotlinx.collections.immutable.persistentSetOf
import kotlinx.collections.immutable.toPersistentSet

private typealias Meet<A> = PersistentSet<A>
private typealias JoinOfMeets<A> = PersistentSet<Meet<A>>

/**
 * The free distributive lattice over an arbitrary set `A` of elements. In addition to lattice
 * identities, the following hold:
 *
 * `a /\ (b \/ c) == (a /\ b) \/ (a /\ c)`
 *
 * `a \/ (b /\ c) == (a \/ b) /\ (a \/ c)`
 */
class FreeDistributiveLattice<A> private constructor(joinOfMeets: JoinOfMeets<A>) :
    HeytingAlgebra<FreeDistributiveLattice<A>> {
    private val joinOfMeets = removeRedundant(joinOfMeets)

    constructor(element: A) : this(persistentSetOf(persistentSetOf(element)))

    override fun join(that: FreeDistributiveLattice<A>): FreeDistributiveLattice<A> {
        return FreeDistributiveLattice(joinOfMeets.addAll(that.joinOfMeets))
    }

    override fun meet(that: FreeDistributiveLattice<A>): FreeDistributiveLattice<A> {
        var candidates: JoinOfMeets<A> = persistentSetOf()
        for (meet1 in joinOfMeets) {
            for (meet2 in that.joinOfMeets) {
                candidates = candidates.add(meet1.addAll(meet2))
            }
        }
        return FreeDistributiveLattice(candidates)
    }

    /**
     * Returns the relative pseudocomplement of `this` relative to `that`. the relative
     * pseudocomplement is greatest x s.t. `this & x <= that`.
     *
     *
     * How does this work? we are dealing with constraints of the form `(A1 | ... | Am) & x
     * <= B1 | ... | Bn`
     *
     *
     * which can be rewritten as `(A1&x) | ... | (Am&x) <= B1 | ... | Bn`
     *
     *
     * This inequality only holds true if every meet on the left can be "covered" on the right s.t.
     * a meet on the right side is a subset of the meet on the left side. For every meet on the left
     * Ai, we complement it with every meet on the right Bj. because we want the greatest solution, we
     * join these complements together, arriving at an upper bound for x: `x <= Ci1 | ... | Cin`
     *
     *
     * where `Cij = Bj \ Ai`.
     *
     *
     * But we have to do the same process for all meets on the left, so we get m upper bounds.
     * these have to be all simultaneously satisfied, so we take the meet of the upper bounds: `x = (C11 | ... | C1n) & ... & (Cm1 | ... | Cmn)`
     *
     *
     * The algorithm below computes exactly this solution.
     */
    override fun imply(that: FreeDistributiveLattice<A>): FreeDistributiveLattice<A> {
        var result = bounds<A>().top
        joinOfMeets.forEach { thisMeet ->
            val newJoinOfMeets =
                that.joinOfMeets.map { thatMeet -> thatMeet.removeAll(thisMeet) }.toPersistentSet()
            result = result.meet(FreeDistributiveLattice(newJoinOfMeets))
        }
        return result
    }

    override fun equals(other: Any?): Boolean =
        other is FreeDistributiveLattice<*> && this.joinOfMeets == other.joinOfMeets

    override fun hashCode(): Int =
        joinOfMeets.hashCode()

    override fun toString(): String {
        fun meetToString(meet: Meet<A>): String {
            val elements = meet.map { it.toString() }.sorted()
            val body = elements.joinToString(" \u2227 ")
            return if (meet.size > 1) "($body)" else body
        }

        return when (this) {
            bounds<A>().top ->
                "\u22A4"

            bounds<A>().bottom ->
                "\u22A5"

            else -> {
                val meets = joinOfMeets.map { meetToString(it) }.sorted()
                val body = meets.joinToString(" \u2228 ")
                if (joinOfMeets.size > 1) "($body)" else body
            }
        }
    }

    companion object {
        private object Bounds : BoundedLattice<FreeDistributiveLattice<Nothing>> {
            override val bottom: FreeDistributiveLattice<Nothing> =
                FreeDistributiveLattice(persistentSetOf())

            override val top: FreeDistributiveLattice<Nothing> =
                FreeDistributiveLattice(persistentSetOf(persistentSetOf()))
        }

        @Suppress("UNCHECKED_CAST")
        fun <A> bounds(): BoundedLattice<FreeDistributiveLattice<A>> =
            Bounds as BoundedLattice<FreeDistributiveLattice<A>>

        /** Remove redundant meets according to [isRedundant]. */
        private fun <A> removeRedundant(joinOfMeets: JoinOfMeets<A>): JoinOfMeets<A> {
            return joinOfMeets.filter { meet -> !isRedundant(joinOfMeets, meet) }.toPersistentSet()
        }

        /**
         * Given `m_1 \/ m_2 \/ ... \/ m_n`, if any `m_i` is a strict subset of `m_j`,
         * then `m_j` is redundant.
         */
        private fun <A> isRedundant(joinOfMeets: JoinOfMeets<A>, j: Meet<A>): Boolean =
            joinOfMeets.any { it != j && j.containsAll(it) }
    }
}
